modules/notifications/notifications.php

```diff
@@ -4,1292 +4,1453 @@
-		$message_headers = apply_filters( 'ef_notification_send_email_message_headers', $message_headers, $action, $post );
-		
-		if( EF_NOTIFICATION_USE_CRON ) {
-			$this->schedule_emails( $recipients, $subject, $message, $message_headers );
-		} else if ( !empty( $recipients ) ) {
-			foreach( $recipients as $recipient ) {
-				$this->send_single_email( $recipient, $subject, $message, $message_headers );
+
+		/**
+		 * send_email()
+		 */
+		public function send_email( $action, $post, $subject, $message, $message_headers = '' ) {
+
+			// Get list of email recipients -- set them CC
+			$recipients = $this->_get_notification_recipients( $post, true );
+
+			if ( $recipients && ! is_array( $recipients ) ) {
+				$recipients = explode( ',', $recipients );
+			}
+
+			$subject         = apply_filters( 'ef_notification_send_email_subject', $subject, $action, $post );
+			$message         = apply_filters( 'ef_notification_send_email_message', $message, $action, $post );
+			$message_headers = apply_filters( 'ef_notification_send_email_message_headers', $message_headers, $action, $post );
+
+			if ( EF_NOTIFICATION_USE_CRON ) {
+				$this->schedule_emails( $recipients, $subject, $message, $message_headers );
+			} elseif ( ! empty( $recipients ) ) {
+				foreach ( $recipients as $recipient ) {
+					$this->send_single_email( $recipient, $subject, $message, $message_headers );
+				}
 			}
 		}
-	}
-	
-	/**
-	 * Schedules emails to be sent in succession
-	 * 
-	 * @param mixed $recipients Individual email or array of emails
-	 * @param string $subject Subject of the email
-	 * @param string $message Body of the email
-	 * @param string $message_headers. (optional) Message headers
-	 * @param int $time_offset (optional) Delay in seconds per email
-	 */
-	function schedule_emails( $recipients, $subject, $message, $message_headers = '', $time_offset = 1 ) {
-		$recipients = (array) $recipients;
-		
-		$send_time = time();
-		
-		foreach( $recipients as $recipient ) {
-			wp_schedule_single_event( $send_time, 'ef_send_scheduled_email', array( $recipient, $subject, $message, $message_headers ) );
-			$send_time += $time_offset;
+
+		/**
+		 * Send notifications to Slack
+		 */
+		public function send_to_webhook( $message ) {
+			$webhook_url = $this->module->options->webhook_url;
+
+			// Bail if the webhook URL is not set
+			if ( empty( $webhook_url ) ) {
+				return;
+			}
+
+			// apply filters to the message
+			$message = apply_filters( 'ef_notification_send_to_webhook_message', $message );
+
+			// Set up the payload
+			$payload = array(
+				'text' => $message,
+			);
+
+			// apply filters to the payload
+			$payload = apply_filters( 'ef_notification_send_to_webhook_payload', $payload );
+
+			// Send the notification
+			$response = wp_remote_post(
+				$webhook_url,
+				array(
+					'body'    => wp_json_encode( $payload ),
+					'headers' => array( 'Content-Type' => 'application/json' ),
+				)
+			);
+			if ( is_wp_error( $response ) ) {
```

Comment:
I tried to test the error pathway by giving an invalid webhook url, and the commenting system just hangs as a result. It seems like it's not able to parse the ajax response, and as a result it just gets stuck on an endless spinner. This seems like it's happening in the editorial-comments.js.

Reply:
In addition, should we print the ajax response? Wondering if that would reveal some sensitive info, and instead if we should just say that "Sending to webhook failed" and instead send some extra debugging info to the error logs

Reply:
I've fixed the error handling to not hang the UI in 57a6806a6e8761cf71e06404c7d05c2728f86731 but the error display is kinda terrible. I've tried to style it as best as I can but I'm still no happy with it.

![Screenshot 2024-06-19 at 2 59 10 PM](https://github.com/Automattic/Edit-Flow/assets/1227524/b50be03f-9d81-4422-b83c-76f3e45cae5a)

Reply:
What if we ran a check when the webhook was saved, so we could show the error message then and there? That way, it's easy to prevent a save if the webhook is invalid and we could change this to something like "Unable to send notification to webhook provided"

Reply:
I'm not sure what you mean by that @ingeniumed, can you give an example?

Reply:
So what I'm suggesting is two fold:

- When the slack webhook url is provided, and the save button is pressed we run a test on that url. We post a test message that says "Integrated successfully with EditFlow". If this doesn't happen, the save fails then and there. This will reduce the chances of failures happening down the line (doesn't eliminate it but sure helps).
- When a notification fails to be sent in the block editor,  we change the error that's shown to be more specific i.e. "Unable to send notification to webhook provided".

Hope that clears up what I mean.

Reply:
Gotcha. I get it now 👍